src/dcos_e2e/docker_utils/__init__.py

```diff
@@ -0,0 +1,141 @@
+"""
+Helpers for creating loopback devices on Docker.
+"""
+
+from typing import Any, Dict, Optional, Tuple
+
+import docker
+
+
+class DockerLoopbackVolume():
+    """
+    A loopback device sidecar, created in a Docker container.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        size: int,
+        labels: Optional[Dict[str, str]] = None,
+    ) -> None:
+        """
+        Create a loopback device pointing to a block device running
+        in a container. This can be used to provide an unformatted
+        disk to a cluster.
+
+        Args:
+            name: Unique name of the loopback volume.
+            size: Size of the block device in Megabytes.
+            labels: Docker labels to add to the container.
+        """
+        self.size = size
+
+        self.container, self.path = DockerLoopbackVolume.create(
+            name=name,
+            size=size,
+            labels=labels,
+        )
+
+    @staticmethod
+    def create(
+        name: str,
+        size: int,
+        labels: Optional[Dict[str, str]] = None,
+    ) -> Tuple[docker.models.containers.Container, str]:
+        """
+        Create a loopback device pointing to a block device running
+        in a container.
+
+        Args:
+            name: Unique name of the loopback volume.
+            size: Size of the block device in Megabytes.
+            labels: Docker labels to add to the container.
+
+        Returns:
+            A tuple containing the container and the path of
+            the loopback device.
+        """
+        client = docker.from_env(version='auto')
+
+        # We use CentOS 7 here, as it provides all the binaries we need
+        # and might already be pulled as it is a distribution supported
+        # by DC/OS.
+        container = client.containers.create(
+            name='dcos-e2e-loopback-sidecar-{name}'.format(name=name),
+            privileged=True,
+            detach=True,
+            tty=True,
+            image='centos:7',
+            labels=labels or {},
+        )
+        container.start()
+
+        create_loopback_device = (
+            'dd if=/dev/zero of=/volume0 bs=1M count={size};'
+        ).format(size=size)
+
+        exit_code, output = container.exec_run(
+            cmd=['/bin/bash', '-c', create_loopback_device],
+        )
+        assert exit_code == 0, output.decode()
+
+        setup_loopback_device = 'losetup --find --show /volume0;'
```

Comment:
Future work: `losetup` is unreliable due to race conditions in the lo driver itself. I don't think this has yet been patched yet on the RHEL/CentOS kernels. If the Docker's host OS has a recent enough kernel then this won't be a problem. This is not the case, for example, w/ CI testing for the dcos-storage project. Hence, dcos-storage "in-repo" test scripts jump through a few hoops to "validate" the loopback device before trying to use it - this avoids flakes both in local development and on CI.